File: web/src/app/services/random-table.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { RandomTable, RandomTableCreate } from './random-table.model';

/**
 * CRUD des tables aléatoires (campagne). Mirroir de NpcService.
 */
@Injectable({ providedIn: 'root' })
export class RandomTableService {
  private apiUrl = '/api/random-tables';

  constructor(private http: HttpClient) {}

  getByCampaign(campaignId: string): Observable<RandomTable[]> {
    return this.http.get<RandomTable[]>(`${this.apiUrl}/campaign/${campaignId}`);
  }

  getById(id: string): Observable<RandomTable> {
    return this.http.get<RandomTable>(`${this.apiUrl}/${id}`);
  }

  create(payload: RandomTableCreate): Observable<RandomTable> {
    return this.http.post<RandomTable>(this.apiUrl, payload);
  }

  update(id: string, payload: RandomTable): Observable<RandomTable> {
    return this.http.put<RandomTable>(`${this.apiUrl}/${id}`, payload);
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  /** Réordonne les tables aléatoires d'une campagne : order = position. */
  reorder(orderedIds: string[]): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/reorder`, { orderedIds });
  }

  /** Recherche par nom — alimente la recherche globale (Ctrl+K). */
  search(q: string): Observable<RandomTable[]> {
    return this.http.get<RandomTable[]>(`${this.apiUrl}/search`, { params: { q } });
  }

  /** Génère une PROPOSITION de table via l'IA (non persistée) à préremplir dans le formulaire. */
  generate(campaignId: string, description: string, diceFormula: string): Observable<RandomTable> {
    return this.http.post<RandomTable>(`${this.apiUrl}/generate`, { campaignId, description, diceFormula });
  }

  /** Improvisation IA d'un court récit sur un résultat tiré (en partie). */
  improvise(
    campaignId: string, tableName: string, resultLabel: string, resultDetail: string
  ): Observable<{ narration: string }> {
    return this.http.post<{ narration: string }>(
      `${this.apiUrl}/improvise`,
      { campaignId, tableName, resultLabel, resultDetail }
    );
  }
}

File: core/src/main/java/com/loremind/application/campaigncontext/RandomTableService.java
package com.loremind.application.campaigncontext;

import com.loremind.domain.campaigncontext.randomtable.RandomTable;
import com.loremind.domain.shared.ReorderSupport;
import com.loremind.domain.campaigncontext.randomtable.RandomTableEntry;
import com.loremind.domain.campaigncontext.ports.RandomTableGenerator;
import com.loremind.domain.campaigncontext.ports.RandomTableRepository;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Service d'application pour les tables aléatoires (campagne).
 */
@Service
public class RandomTableService {

    private final RandomTableRepository repository;
    private final RandomTableGenerator generator;
    private final CampaignContextFormatter campaignContextFormatter;

    public RandomTableService(
            RandomTableRepository repository,
            RandomTableGenerator generator,
            CampaignContextFormatter campaignContextFormatter) {
        this.repository = repository;
        this.generator = generator;
        this.campaignContextFormatter = campaignContextFormatter;
    }

    public record TableData(
            String name,
            String description,
            String diceFormula,
            String icon,
            List<RandomTableEntry> entries,
            String campaignId,
            Integer order
    ) {}

    public RandomTable createTable(TableData data) {
        int order = data.order() != null ? data.order() : nextOrderFor(data.campaignId());
        RandomTable table = RandomTable.builder()
                .name(data.name())
                .description(data.description())
                .diceFormula(data.diceFormula())
                .icon(data.icon())
                .entries(copyEntries(data.entries()))
                .campaignId(data.campaignId())
                .order(order)
                .build();
        return repository.save(table);
    }

    public Optional<RandomTable> getTableById(String id) {
        return repository.findById(id);
    }

    public List<RandomTable> getTablesByCampaignId(String campaignId) {
        return repository.findByCampaignId(campaignId);
    }

    public RandomTable updateTable(String id, TableData data) {
        RandomTable existing = repository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Table aléatoire introuvable: " + id));
        existing.setName(data.name());
        existing.setDescription(data.description());
        existing.setDiceFormula(data.diceFormula());
        existing.setIcon(data.icon());
        existing.setEntries(copyEntries(data.entries()));
        if (data.order() != null) {
            existing.setOrder(data.order());
        }
        return repository.save(existing);
    }

    public void deleteTable(String id) {
        repository.deleteById(id);
    }

    /** Réordonne les tables aléatoires d'une campagne : {@code order} = position. */
    @org.springframework.transaction.annotation.Transactional
    public void reorderTables(List<String> orderedIds) {
        ReorderSupport.reorder(orderedIds,
                repository::findById,
                RandomTable::setOrder,
                repository::save);
    }

    public List<RandomTable> searchTables(String query) {
        if (query == null || query.isBlank()) return List.of();
        return repository.searchByName(query.trim());
    }

    /** Génère une PROPOSITION de table (non persistée) via l'IA, contextualisée campagne. */
    public RandomTable generateProposal(String campaignId, String description, String diceFormula) {
        String formula = (diceFormula == null || diceFormula.isBlank()) ? "1d20" : diceFormula;
        RandomTableGenerator.GeneratedTable g = generator.generate(
                description, formula, campaignContextFormatter.format(campaignId));
        return RandomTable.builder()
                .name(g.name())
                .description(g.description())
                .diceFormula(formula)
                .campaignId(campaignId)
                .entries(g.entries() != null ? new ArrayList<>(g.entries()) : new ArrayList<>())
                .build();
    }

    /** Brode un court récit IA sur un résultat tiré (pour la partie). */
    public String improviseRoll(String campaignId, String tableName, String resultLabel, String resultDetail) {
        return generator.improvise(tableName, resultLabel, resultDetail, campaignContextFormatter.format(campaignId));
    }

    private static List<RandomTableEntry> copyEntries(List<RandomTableEntry> entries) {
        return entries != null ? new ArrayList<>(entries) : new ArrayList<>();
    }

    private int nextOrderFor(String campaignId) {
        return repository.findByCampaignId(campaignId).stream()
                .mapToInt(RandomTable::getOrder)
                .max()
                .orElse(-1) + 1;
    }
}
